File: src/main/java/com/bookislife/jgsk/java/_05_datatype/DataType.java
package com.bookislife.jgsk.java._05_datatype;

import java.math.BigDecimal;

/**
 * Created by SidneyXu on 2015/09/16.
 */
public class DataType {

    public static void main(String[] args) {
        //  Data Types
        int x = 2;
        float y = 0.1f;
        boolean flag = false;

        //  BigDecimal
        System.out.println(2.0 - 1.8);  //  0.19999999999999996
        System.out.println(new BigDecimal(2.0).subtract(new BigDecimal(1.8)));  //  0.19999999999999995559...
        System.out.println(new BigDecimal("2.0").subtract(new BigDecimal("1.8")));  //  0.2

        //  Type Conversion
        //  Implicit Conversion
        byte b = 1;
        int i = b;
        char c = 1;

        //  Explicit Conversion
        System.out.println((int) 99.98);
        System.out.println((char) 99);
        System.out.println(Double.valueOf("99.12"));

        //  Type Inference
        Object foo = "foo";
        bar(foo);
    }

    private static void bar(Object foo) {
        if (foo instanceof String) {
            String fooString = (String) foo;
            System.out.println(fooString.toUpperCase());
        }
    }
}

File: src/main/kotlin/com/bookislife/jgsk/kotlin/_05_datatype/DataType.kt
package com.bookislife.jgsk.kotlin._05_datatype

import java.math.BigDecimal

/**
 * Created by SidneyXu on 2015/09/16.
 */
fun main(args: Array<String>) {
    //  Everything is Object
    1.toString()

    //  Data Types
    var x: Int = 2
    var y: Float = 0.1f
    var flag: Boolean = false
    var z = 1
    //  Wrong!!
    //    z = false

    //  BigDecimal
    println(BigDecimal("2").subtract(BigDecimal("1.8")));

    //  Type Conversion
    var b: Byte = 1
    //  Wrong!! implicit conversion not supported
    //    val i: Int = b
    //  Wrong!! Char is not a number
    //    val c: Char = 1

    //  Explicit Conversion
    println(99.98.toInt())
    println(99.toChar())
    println("99.12".toDouble())

    //  Type Inference
    val foo = "foo"
    val fooString: String = foo as String
    bar(foo)
}

fun bar(foo: Any) {
    if (foo is String) {
        println(foo.toUpperCase())
    }
}